mon-pix/app/routes/campaigns/start-or-resume.js

```diff
@@ -6,62 +7,148 @@ export default class StartOrResumeRoute extends Route.extend(SecuredRouteMixin)
   @service currentUser;
   @service session;
 
-  campaignCode = null;
-  participantExternalId = null;
-  campaignIsRestricted = false;
-  userHasSeenLanding = false;
-  authenticationRoute = 'inscription';
+  state = null;
+
+  constructor() {
+    super(...arguments);
+    this._resetState();
+  }
 
   beforeModel(transition) {
-    this.campaignCode = transition.to.parent.params.campaign_code;
-    this.participantExternalId = transition.to.queryParams.participantExternalId;
-    this.campaignIsRestricted = transition.to.queryParams.campaignIsRestricted;
-    this.userHasSeenLanding = transition.to.queryParams.hasSeenLanding;
+    this.authenticationRoute = 'inscription';
+    const campaign = this.modelFor('campaigns');
+    this._updateStateFrom({ campaign, queryParams: transition.to.queryParams });
 
-    if (this._userIsUnauthenticated() && this.campaignIsRestricted) {
+    if (this._shouldLoginToAccessRestrictedCampaign) {
       this.authenticationRoute = 'campaigns.restricted.login-or-register-to-access';
+      super.beforeModel(...arguments); // Force logging
+      return;
+    }
+
+    if (this._shouldJoinRestrictedCampaign) {
+      if (this.currentUser.user.mustValidateTermsOfService) {
+        super.beforeModel(...arguments); // Force accepting terms of service
```

Comment:
Ici je vois pas comment ça redirige vers les terms of service 🤔  Quelqu'un peut m'expliquer ?

Reply:
Faut reconnaître que c'est pas limpide. Mais il me semble que ça marche grâce à cette ligne au-dessus :
```js
this.authenticationRoute = 'inscription';
```

Reply:
Le `super.beforeModel()` ici (et au dessus) va plutôt d'abord appeler celui du `SecuredRouteMixin`. `this.authenticationRoute = 'inscription';` permet de surcharger la valeur par défaut (route login) a appelé si l'utilisateur n'est pas authentifié.

Reply:
Je crois que c'est a cause de ça : 

```js
export default class StartOrResumeRoute extends Route.extend(SecuredRouteMixin) {
```
et
mon-pix/app/mixins/secured-route-mixin.js
```js
 beforeModel(transition) {
    if (!this.session.get('isAuthenticated')) {
      this.session.set('attemptedTransition', transition);
      this.router.transitionTo(this.authenticationRoute);
    } else if (this.currentUser.user.mustValidateTermsOfService) {
      this.session.set('attemptedTransition', transition);
      this.router.transitionTo('terms-of-service');
    } else {
      return this._super(...arguments);
    }
}
```

C'est le else if qui fait que tu retourne voir les terms of services

Reply:
J'ai encore du mal à bien comprendre ce que fait ce code de la fonction beforeModel.

- Les appels à `super.beforeModel` masque un comportement complexe pas forcément facile à appréhender.
(Si j'ai compris  : Un appelle à `super.beforeModel`== tu ne peux pas accéder à la campagne toute suite ?)
- le `this.authenticationRoute` qui n'est utile que dans certain cas où on appelle `super.beforeModel` mais pas tous.
- La présence des commentaires pour moi montre que l'intention du code n'est pas claire

Je ne suis pas que l'utilisation de l'héritage (avec la classe `SecuredRouteMixin`) soit le meilleur choix ici.
Ça rend le code moins clair et moins compréhensible (pour moi) du coup je challengerais bien l'utilisation de l'héritage dans ce cas.